### PR TITLE
CFE-956: [v2] feat: adds tags on the related images

### PR DESCRIPTION
**Pull request by aguidirh**
# Description

Before this implementation the related images were being stored only by digest. Now the related images are going to have a tag that is the 12 first chars of its digest.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

```
apiVersion: mirror.openshift.io/v1alpha2
kind: ImageSetConfiguration
mirror:
  operators:
    - catalog: registry.redhat.io/redhat/redhat-operator-index:v4.12
      packages:
        - name: aws-load-balancer-operator
  additionalImages: 
    - name: registry.redhat.io/ubi8/ubi
    - name: registry.redhat.io/ubi9/ubi
```

MirrorToDisk
```
./bin/oc-mirror -c /home/aguidi/go/src/github.com/aguidirh/oc-mirror/alex-isc/isc.yaml file:///home/aguidi/go/src/github.com/aguidirh/oc-mirror/CFE-956-v2 --v2
```

DiskToMirror
```
./bin/oc-mirror -c /home/aguidi/go/src/github.com/aguidirh/oc-mirror/alex-isc/isc-additional-image.yaml --from /home/aguidi/go/src/github.com/aguidirh/oc-mirror/CFE-956-v2 docker://localhost:9000 --v2
```

One example of tags with the first 12 chars of the digest:
```
[aguidi@fedora oc-mirror]$ curl -X GET http://localhost:9000/v2/registry.redhat.io/albo/aws-load-balancer-controller-rhel8/tags/list | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   124  100   124    0     0  25218      0 --:--:-- --:--:-- --:--:-- 31000
{
  "name": "registry.redhat.io/albo/aws-load-balancer-controller-rhel8",
  "tags": [
    "d7bc36451217",
    "cfd210a2cb70",
    "cad8f6380b4d"
  ]
}
```

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: v2/pkg/additional/new.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+
+func New(log clog.PluggableLoggerInterface,
+		config v1alpha2.ImageSetConfiguration,
+		opts mirror.CopyOptions,
+		mirror mirror.MirrorInterface,
+		manifest manifest.ManifestInterface,
+		localStorageFQDN string,
+) CollectorInterface {
+	if localStorageFQDN != "" {
```
Comment (lmzuccarelli) [inline v2/pkg/additional/new.go:18]: @aguidirh - this is good, with simple logic we can plugin different implementations

File: v2/pkg/additional/local_stored_collector.go [hunk 2]
```diff
@@ -76,7 +75,12 @@ func (o *LocalStorageCollector) AdditionalImagesCollector(ctx context.Context) (
...
 				src = dockerProtocol + strings.Join([]string{o.LocalStorageFQDN, img.Name}, "/")
-				dst = strings.Join([]string{o.Opts.Destination, img.Name}, "/")
+
+				if isDigestName(img.Name) {
+					dest = strings.Join([]string{o.Opts.Destination, imageShortName(img.Name) + ":" + imageHash(img.Name)[:hashTruncLen]}, "/")
+				} else {
+					dest = strings.Join([]string{o.Opts.Destination, img.Name}, "/")
+				}
 
 				// the following is for having the destination without the initial domain name => later
```
Comment (lmzuccarelli) [inline v2/pkg/additional/local_stored_collector.go:85]: @aguidirh - I know this was not part of the PR but could you remove all unwanted comments

File: v2/pkg/cli/executor.go [hunk 3]
```diff
@@ -330,14 +331,15 @@ func (o *ExecutorSchema) Complete(args []string) {
...
+	o.LocalStorageFQDN = "localhost:" + strconv.Itoa(int(o.Opts.Global.Port))
 
 	client, _ := release.NewOCPClient(uuid.New())
 
 	signature := release.NewSignatureClient(o.Log, &o.Config, &o.Opts)
 	cn := release.NewCincinnati(o.Log, &o.Config, &o.Opts, client, false, signature)
 	o.Release = release.New(o.Log, o.Config, o.Opts, o.Mirror, o.Manifest, cn)
-	o.Operator = operator.NewWithLocalStorage(o.Log, o.Config, o.Opts, o.Mirror, o.Manifest, "localhost:"+strconv.Itoa(int(o.Opts.Global.Port)))
-	o.AdditionalImages = additional.NewWithLocalStorage(o.Log, o.Config, o.Opts, o.Mirror, o.Manifest, "localhost:"+strconv.Itoa(int(o.Opts.Global.Port)))
+	o.Operator = operator.New(o.Log, o.Config, o.Opts, o.Mirror, o.Manifest, o.LocalStorageFQDN)
```
Comment (lmzuccarelli) [inline v2/pkg/cli/executor.go:341]: @aguidirh -  much simpler !!!

File: v2/pkg/operator/local_stored_collector.go [hunk 4]
```diff
@@ -208,7 +196,12 @@ func (o *LocalStorageCollector) prepareD2MCopyBatch(log clog.PluggableLoggerInte
...
 			src = dockerProtocol + strings.Join([]string{o.LocalStorageFQDN, img.Image}, "/")
-			dst = strings.Join([]string{o.Opts.Destination, img.Image}, "/")
+
+			if isDigestName(img.Image) {
+				dest = strings.Join([]string{o.Opts.Destination, imageShortName(img.Image) + ":" + imageHash(img.Image)[:hashTruncLen]}, "/")
+			} else {
+				dest = strings.Join([]string{o.Opts.Destination, img.Image}, "/")
+			}
 
 			// the following is for having the destination without the initial domain name => later
```
Comment (lmzuccarelli) [inline v2/pkg/operator/local_stored_collector.go:206]: @aguidirh - same here i.e can you remove all  comments the hide old code ?

File: v2/pkg/operator/local_stored_collector.go [hunk 5]
```diff
@@ -237,15 +230,21 @@ func (o *LocalStorageCollector) prepareM2DCopyBatch(log clog.PluggableLoggerInte
 	for _, relatedImgs := range images {
 		for _, img := range relatedImgs {
 			imgName := img.Image
-			src := ""
+			var src string
+			var dest string
 			if !strings.Contains(src, "://") { // no transport was provided, assume docker://
```
Comment (lmzuccarelli) [inline v2/pkg/operator/local_stored_collector.go:235]: nit: could you put the comment above - its just a matter of style

**Review by lmzuccarelli — COMMENTED** [review]
great work Alex

File: v2/pkg/operator/local_stored_collector.go (hunk 5 above)
Comment (aguidirh) [reply v2/pkg/operator/local_stored_collector.go]: Done

File: v2/pkg/additional/new.go (hunk 1 above)
Comment (aguidirh) [reply v2/pkg/additional/new.go:18]: Thanks!

File: v2/pkg/additional/new.go [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+		return &LocalStorageCollector{Log: log, Config: config, Opts: opts, Mirror: mirror, Manifest: manifest, LocalStorageFQDN: localStorageFQDN}
+	} else {
+		return &Collector{Log: log, Config: config, Opts: opts, Mirror: mirror, Manifest: manifest}
+	}
```
Comment (sherine-k) [inline v2/pkg/additional/new.go:22]: Nice! 
Just one thing I thought to point out is that this means we need to correctly implement, test, maintain both implementations.
For the moment, I'm not sure we have done that.

File: v2/pkg/additional/local_stored_collector.go (hunk 2 above)
Comment (sherine-k) [reply v2/pkg/additional/local_stored_collector.go:87]: These will come in handy when we remove the registry from the destination image url
today, we are copying registry.redhat.io/ubi8/ubi to localhost:5000/registry.redhat.io/ubi8/ubi

File: v2/pkg/operator/local_stored_collector.go [hunk 7]
```diff
@@ -254,3 +255,27 @@ func (o *LocalStorageCollector) prepareM2DCopyBatch(log clog.PluggableLoggerInte
 	}
 	return result, nil
 }
+
+func isDigestName(img string) bool {
```
Comment (sherine-k) [inline v2/pkg/operator/local_stored_collector.go:259]:
```suggestion
func isImageByDigest(img string) bool {
```

File: v2/pkg/operator/local_stored_collector.go [hunk 8]
```diff
@@ -254,3 +255,27 @@ func (o *LocalStorageCollector) prepareM2DCopyBatch(log clog.PluggableLoggerInte
...
+	return strings.Contains(img, "@")
+}
+
+func imageShortName(img string) string {
```
Comment (sherine-k) [inline v2/pkg/operator/local_stored_collector.go:263]:
```suggestion
func imageName(reference string) string {
```
Not sure if this suggested name is better tbh...
According to https://github.com/containers/image/blob/main/docker/reference/reference.go what we are extracting is the name from the reference.

**Review by sherine-k — COMMENTED** [review]
Thanks Alex, 
This clarifies the code by much and fixes this issue we've been having with pulling the images!

File: v2/pkg/operator/local_stored_collector.go (hunk 7 above)
Comment (aguidirh) [reply v2/pkg/operator/local_stored_collector.go]: Good suggestion, done!

File: v2/pkg/operator/local_stored_collector.go (hunk 8 above)
Comment (aguidirh) [reply v2/pkg/operator/local_stored_collector.go]: Good suggestion, done!

File: v2/pkg/additional/new.go (hunk 6 above)
Comment (aguidirh) [reply v2/pkg/additional/new.go:22]: Probably we are not going to maintain both implementations. But I agree with you that we need to create tests for the implementation we are going to maintain.
